jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java

```diff
@@ -85,12 +86,20 @@ protected HttpClient getHttpClient()
     
     protected List<HeaderInfo> getHeaderInfo(String value) throws IllegalArgumentException
     {
+        return getHeaderInfo(value, 0);
+    }
+    
+    protected List<HeaderInfo> getHeaderInfo(String value, int depth) throws IllegalArgumentException
+    {
+        if(depth > MAX_DEPTH)
+            throw new IllegalStateException("too many challanges");
```

Comment:
Uh, no. Is there a reason you cannot iterate rather than recurse ?

Reply:
@lachlan-roberts, @gregw this is an iterative solution that seems to work for me, can you please check if it does for you:

```java
List<String> challenges = new ArrayList<>();
String fullChallenge = "Basic a=1,b=2,Negotiate c=2,Scheme d=3,Foo";
String challenge = fullChallenge;
while (true)
{
    Matcher matcher = MULTIPLE_CHALLENGE_PATTERN.matcher(challenge);
    if (matcher.matches())
    {
        challenge = matcher.group(1);
        challenges.add(0, matcher.group(2));
    }
    else
    {
        challenges.add(0, challenge);
        break;
    }
}
System.err.println("challenges = " + challenges);
```

Reply:
I had discounted that approach as creating the matcher every iteration... but then I guess this is moderately rare, so perhaps it's not such a problem...

Reply:
@gregw the matcher is created as well for every recursion, so the approach is equivalent allocation-wise, but iteration is less better.
The only gripe is that the list would need to be reversed at the end, or - like above - the challenges need to be inserted at index `0`. Perhaps a single `Collections.reverse(challenges)` at the end is better.